x-pack/plugin/ccr/src/test/java/org/elasticsearch/xpack/ccr/index/engine/FollowingEngineTests.java

```diff
@@ -278,49 +280,223 @@ private FollowingEngine createEngine(Store store, EngineConfig config) throws IO
+        IndexMetaData leaderIndexMetaData = IndexMetaData.builder(index.getName()).settings(leaderSettings).build();
+        IndexSettings leaderIndexSettings = new IndexSettings(leaderIndexMetaData, leaderSettings);
+        try (Store leaderStore = createStore(shardId, leaderIndexSettings, newDirectory())) {
+            leaderStore.createEmpty();
+            EngineConfig leaderConfig = engineConfig(shardId, leaderIndexSettings, threadPool, leaderStore, logger, xContentRegistry());
+            leaderStore.associateIndexWithNewTranslog(Translog.createEmptyTranslog(
+                leaderConfig.getTranslogConfig().getTranslogPath(), SequenceNumbers.NO_OPS_PERFORMED, shardId, 1L));
+            try (InternalEngine leaderEngine = new InternalEngine(leaderConfig)) {
+                leaderEngine.initializeMaxSeqNoOfUpdatesOrDeletes();
+                leaderEngine.skipTranslogRecovery();
+                Settings followerSettings = Settings.builder()
+                    .put("index.number_of_shards", 1).put("index.number_of_replicas", 0)
+                    .put("index.version.created", Version.CURRENT).put("index.xpack.ccr.following_index", true).build();
+                IndexMetaData followerIndexMetaData = IndexMetaData.builder(index.getName()).settings(followerSettings).build();
+                IndexSettings followerIndexSettings = new IndexSettings(followerIndexMetaData, leaderSettings);
+                try (Store followerStore = createStore(shardId, followerIndexSettings, newDirectory())) {
+                    EngineConfig followerConfig = engineConfig(
+                        shardId, followerIndexSettings, threadPool, followerStore, logger, xContentRegistry());
+                    try (FollowingEngine followingEngine = createEngine(followerStore, followerConfig)) {
+                        wrappedTask.accept(leaderEngine, followingEngine);
+                    }
+                }
+            }
         }
-        return new Engine.Index(
-            new Term("_id", parsedDocument.id()),
-            parsedDocument,
-            seqNo,
-            primaryTerm.get(),
-            version,
-            VersionType.EXTERNAL,
-            origin,
-            System.currentTimeMillis(),
-            autoGeneratedIdTimestamp,
-            randomBoolean());
     }
 
+    private void fetchOperations(AtomicBoolean stopped, AtomicLong lastFetchedSeqNo,
+                                 InternalEngine leader, FollowingEngine follower) throws IOException {
+        final MapperService mapperService = EngineTestCase.createMapperService("test");
+        final TranslogHandler translogHandler = new TranslogHandler(xContentRegistry(), follower.config().getIndexSettings());
+        while (stopped.get() == false) {
+            final long checkpoint = leader.getLocalCheckpoint();
+            final long lastSeqNo = lastFetchedSeqNo.get();
+            if (lastSeqNo < checkpoint) {
```

Comment:
maybe add retries here? (now or later)

Reply:
I have a "retry" logic by delivering some operations multiple times (https://github.com/elastic/elasticsearch/pull/34099/files/a81d1df34965a1781061307dce53d2106f637e3c#diff-103d04fa7060f91d3c4d1919cd0370c1R491). Are you okay with this?

Reply:
We should have out of order delivery because we use multiple threads to fetch operations. However, I pushed https://github.com/elastic/elasticsearch/pull/34099/commits/be86b4f0a7a283cf56bfd041812aabba47fdce46 to shuffle the snapshot. Let me know if you're good with this. I am happy to continue iterating on this.

Reply:
I'm good with this. Thank you.